libraries/botbuilder/src/setSpeakMiddleware.ts

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { ActivityTypes, Channels, Middleware, TurnContext } from 'botbuilder-core';
+import { parseDocument } from 'htmlparser2';
+import { tests } from 'botbuilder-stdlib';
+
+const supportedChannels = new Set([Channels.DirectlineSpeech, Channels.Emulator, 'telephony']);
+
+// Iterate through `obj` and all children in an attempt to locale a key `tag`
+function hasTag(tag: string, nodes: unknown[]): boolean {
+    while (nodes.length) {
+        const item = nodes.shift();
+        console.log(item, tag);
+
+        if (tests.isDictionary(item)) {
+            if (item.tagName === tag) {
+                return true;
+            }
+
+            if (tests.isArray(item.children)) {
+                nodes.push(...item.children);
+            }
+        }
+    }
+
+    return false;
+}
+
+/**
+ * Support the DirectLine speech and telephony channels to ensure the appropriate SSML tags are set on the
+ * Activity Speak property.
+ */
+export class SetSpeakMiddleware implements Middleware {
```

Comment:
I think we might want this in `botbuilder-core`, as the concept _technically_ isn't Node.js dependent. Is `htmlparser2` browser-compatible?

Reply:
Apparently not - according to https://www.npmjs.com/package/htmlparser2 the original source of the fork is, however, this one has orders of magnitude more weekly downloads. I'm inclined to leave it until there is a specific customer use case for using this middleware in a browser bot.

Reply:
That sounds reasonable to me!